server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -134,6 +135,61 @@ public StoredContext stashContext() {
         };
     }
 
+    /**
+     * Removes the specified transient headers from the current context. When the returned
+     * {@link StoredContext} is closed, it will restore these transient headers to their original
+     * value (including restoring them to an <i>unset</i> value if they did not originally exist).
+     * Closing the {@code StoredContext} has no affect on any other header - any headers
+     * (other than those names specified in {@code transientHeadersToStash} that were
+     * added to the {@code ThreadContext} will be retained.
+     *
+     * For example, at the end of the following code, the ThreadContext will have transient
+     * values {@code "a"=1}, {@code "b"=1}, {@code "d"=2} and {@code "c"} will not be set.
+     * <pre>
+     * threadContext.putTransient("a", 1);
+     * threadContext.putTransient("b", 1);
+     * try (ThreadContext.StoredContext restore = threadContext.stashTransientContext(List.of("b", "c")) ) {
+     *   threadContext.putTransient("b", 2);
+     *   threadContext.putTransient("c", 2);
+     *   threadContext.putTransient("d", 2);
+     * }
+     * </pre>
+     */
+    public StoredContext stashTransientHeaders(Collection<String> transientHeadersToStash) {
+        if (transientHeadersToStash.isEmpty()) {
+            // no-op
+            return () -> {
+            };
```

Comment:
I think this should actually make a copy of the struct and restore it, otherwise any additional headers will be leaked back after closing the `StoredContext`.

```suggestion
            final ThreadContextStruct context = threadLocal.get();
            return () -> threadLocal.set(context);
```

Can you also add a test to ensure this? Scenario I am thinking will fail as the code is written:

```java
assertNull(threadContext.getTransient("a");
try (ThreadContext.StoredContext restore = threadContext.stashTransientContext(List.of())) {
    threadContext.putTransient("a", "1");
}
assertNull(threadContext.getTransient("a");
```

Reply:
The leaking is by design. Only the specified headers are restored; for the others, if they're changed inside the stash, they are preserved after restore (think response headers). The renaming from `stashTransientContext` to `stashTransientHeaders` ought to help clarifying this oddity.